File: src/core/application.ts
import { observePath, readPath, writePath } from "./bindings";
import {
  Controller,
  collectControllers,
  controllerFor,
  detachViewTree,
  topViewsIn,
} from "./view";
import { finishActivation, wireSubtree } from "./awakening";
import { Responder, currentFirstResponder, makeFirstResponder, setChainTop } from "./responder";
import { firstResponderFor, syncFirstResponderFromFocus } from "./responder";
import {
  type FragmentHistoryMode,
  type RestorationBindingController,
  type WindowRestoration,
  type CapturedWindowContent,
  type WindowEntry,
  fragmentParams,
  fragmentValue,
  scopedFragmentKey,
  topControllerIn,
  windowContainers,
  writeFragmentParam,
} from "./windows";

/**
 * Cocoa NSApplication analog. App-wide singleton that owns:
 *
 * - The initial wire-pass of the document and the MutationObserver that
 *   keeps later subtree insertions in sync.
 * - Document-level focusin/focusout and keydown/keyup listeners.
 * - Console-debugging hooks (`__controllers`, `__root`, etc.).
 * - The window-template registry and the windows currently on screen.
 *
 * Subclasses can override hook methods. Most apps never touch the class
 * and just call `Application.shared.start(document.body)` at boot.
 */
export class Application extends Responder {
  private static _shared: Application | null = null;
  static get shared(): Application {
    return (this._shared ??= new Application());
  }

  /** Chain terminus — the responder chain bottoms out here (Cocoa
   *  NSApp). No further link. */
  override get nextResponder(): Responder | null {
    return null;
  }

  private rootElement: HTMLElement = document.body;
  private rootControllers: Controller[] = [];
  private windowTemplates = new Map<string, HTMLTemplateElement>();
  private capturedWindowContents = new Map<string, CapturedWindowContent>();
  private windows: WindowEntry[] = [];
  private windowRestoration = new Map<string, WindowRestoration>();
  private applyingFragmentState = false;

  /** One-time startup: install listeners, prepare `[window]`
   * containers, wire their live or templated contents, then apply URL
   * fragment restoration. */
  start(root: HTMLElement = document.body): void {
    Application._shared = this;
    setChainTop(this); // responder chain bottoms out at the Application
    this.rootElement = root;
    this.scanWindowTemplates();
    this.prepareWindowContainers(root);
    this.installKeyRouting();
    this.installFocusTracking();
    this.installFragmentRouting();
    this.watchForChanges(root);

    this.rootControllers = wireSubtree(root, { deferActivation: true });
    this.registerLiveWindows(root, this.rootControllers);
    this.restoreWindowStates();
    this.finishWindowActivation();
    this.installConsoleHooks();
  }

  // ---- windows ----

  /** Show a window from a registered template. Multiple windows can be
   * open simultaneously. Returns a Promise that resolves when the window
   * is dismissed. */
  showWindow(name: string, opts: { into?: HTMLElement } = {}): Promise<void> {
    const container = this.windowContainer(name);
    if (container) {
      return this.loadWindowContent(name, name);
    }

    this.scanWindowTemplates();
    const rootCtrl = this.instantiateTemplate(name, opts.into ?? document.body);
    const node = rootCtrl.view.element;

    if (node instanceof HTMLDialogElement) node.show();

    const savedFR = currentFirstResponder();
    makeFirstResponder(rootCtrl);

    let resolve!: () => void;
    const promise = new Promise<void>((r) => {
      resolve = r;
    });
    this.windows.push({
      root: node,
      controller: rootCtrl,
      name,
      contentName: name,
      savedFirstResponder: savedFR,
      resolve,
    });
    return promise;
  }

  async loadWindowContent(
    windowName: string,
    contentName: string,
    opts: { history?: FragmentHistoryMode } = {},
  ): Promise<void> {
    this.scanWindowTemplates();
    const container = this.windowContainer(windowName);
    if (!container) throw new Error(`Application: no window="${windowName}" container`);

    const old = this.windows.find((w) => w.root === container);
    const previous = currentFirstResponder();

    for (const view of topViewsIn(container)) {
      detachViewTree(view, { removeElement: false });
    }
    container.replaceChildren();

    const fragment = this.cloneWindowContent(contentName);
    container.appendChild(fragment);
    const controllers = wireSubtree(container, { deferActivation: true });
    const controller = topControllerIn(container, controllers);

    const entry: WindowEntry = {
      root: container,
      controller,
      name: windowName,
      contentName,
      savedFirstResponder: old?.savedFirstResponder ?? null,
      resolve: old?.resolve ?? (() => {}),
    };
    const idx = this.windows.indexOf(old!);
    if (idx >= 0) this.windows[idx] = entry;
    else this.windows.push(entry);

    container.setAttribute("name", contentName);
    this.writeFragmentParam(windowName, contentName, opts.history ?? "push");
    this.restoreWindowState(entry, {
      pruneStale: opts.history !== "none",
      writeContent: opts.history !== "none",
    });
    finishActivation(container);
    if (controller) makeFirstResponder(controller);
    else if (previous instanceof Controller && previous.view.element.isConnected) makeFirstResponder(previous);
  }

  private instantiateTemplate(name: string, into: HTMLElement): Controller {
    const tpl = this.windowTemplates.get(name);
    if (!tpl) throw new Error(`Application: no window template "${name}"`);
    const node = tpl.content.firstElementChild?.cloneNode(true) as HTMLElement | null;
    if (!node) throw new Error(`Application: empty window template "${name}"`);

    into.appendChild(node);
    wireSubtree(node);
    const rootCtrl = controllerFor(node);
    if (!rootCtrl) {
      const views = topViewsIn(node);
      if (views.length) {
        for (const view of views) detachViewTree(view);
      }
      node.remove();
      throw new Error(`Application: window "${name}" root has no controller="..." attribute`);
    }
    return rootCtrl;
  }

  /** Tear down the window that owns `controller`. Restores the
   * first-responder that was active when the window was presented. */
  dismiss(controller: Controller): void {
    const idx = this.windows.findIndex((w) => w.controller === controller || w.root.contains(controller.view.element));
    if (idx < 0) return;
    const win = this.windows[idx]!;
    this.windows.splice(idx, 1);

    if (win.root instanceof HTMLDialogElement && win.root.open) win.root.close();
    if (win.controller) detachViewTree(win.controller.view, { removeElement: true });

    const prev = win.savedFirstResponder;
    if (prev instanceof Controller && prev.view.element.isConnected) {
      makeFirstResponder(prev);
    }
    win.resolve();
  }

  private scanWindowTemplates(): void {
    const tpls = document.querySelectorAll<HTMLTemplateElement>('template[for="window"][name], body > template[name]');
    for (const t of tpls) {
      const name = t.getAttribute("name");
      if (name) this.windowTemplates.set(name, t);
    }
  }

  private prepareWindowContainers(root: HTMLElement): void {
    for (const container of windowContainers(root)) {
      const windowName = container.getAttribute("window")?.trim();
      if (!windowName) continue;
      const contentName = container.getAttribute("name")?.trim();
      const hasLiveContent = container.childNodes.length > 0;
      const requestedContent = fragmentParams().get(windowName)?.trim() || null;

      if (contentName && hasLiveContent && !this.windowTemplates.has(contentName)) {
        const fragment = document.createDocumentFragment();
        for (const node of Array.from(container.childNodes)) {
          fragment.appendChild(node.cloneNode(true));
        }
        this.capturedWindowContents.set(contentName, { fragment });
      }

      if (hasLiveContent) {
        if (requestedContent && requestedContent !== contentName) {
          if (this.hasWindowContent(requestedContent)) {
            container.replaceChildren(this.cloneWindowContent(requestedContent));
            container.setAttribute("name", requestedContent);
          } else {
            console.warn(`[Swill] window="${windowName}" requested unknown content "${requestedContent}"`);
          }
        }
        continue;
      }

      const defaultContent = requestedContent || contentName || (this.hasWindowContent(windowName) ? windowName : null);
      if (!defaultContent) continue;
      if (!this.hasWindowContent(defaultContent)) {
        console.warn(`[Swill] window="${windowName}" has no content "${defaultContent}"`);
        continue;
      }
      container.appendChild(this.cloneWindowContent(defaultContent));
      if (contentName == null || requestedContent) container.setAttribute("name", defaultContent);
    }
  }

  private registerLiveWindows(root: HTMLElement, controllers: Controller[]): void {
    for (const container of windowContainers(root)) {
      const name = container.getAttribute("window")?.trim();
      if (!name) continue;
      const controller = topControllerIn(container, controllers);
      this.windows.push({
        root: container,
        controller,
        name,
        contentName: container.getAttribute("name")?.trim() || undefined,
        savedFirstResponder: null,
        resolve: () => {},
      });
    }
  }

  private windowContainer(name: string): HTMLElement | null {
    return this.rootElement.querySelector<HTMLElement>(`[window="${CSS.escape(name)}"]`);
  }

  private hasWindowContent(name: string): boolean {
    return this.windowTemplates.has(name) || this.capturedWindowContents.has(name);
  }

  private cloneWindowContent(name: string): DocumentFragment {
    const tpl = this.windowTemplates.get(name);
    if (tpl) return tpl.content.cloneNode(true) as DocumentFragment;
    const captured = this.capturedWindowContents.get(name);
    if (captured) return captured.fragment.cloneNode(true) as DocumentFragment;
    throw new Error(`Application: no window content template "${name}"`);
  }

  private restoreWindowStates(): void {
    for (const entry of this.windows) {
      this.restoreWindowState(entry, { pruneStale: false, writeContent: true });
    }
  }

  private restoreWindowState(entry: WindowEntry, opts: { pruneStale: boolean; writeContent: boolean }): void {
    if (!entry.name) return;
    const staleKeys = this.disposeWindowRestoration(entry.name);
    const controller = entry.controller as (Controller & RestorationBindingController) | null;
    const bindings = controller?.restorationBindings?.() ?? {};
    const keys = new Set(Object.keys(bindings).map((key) => scopedFragmentKey(entry.name!, key)));

    if (opts.writeContent && entry.contentName) {
      this.writeFragmentParam(entry.name, entry.contentName, "replace");
    }
    if (opts.pruneStale) {
      for (const key of staleKeys) {
        if (!keys.has(key)) this.writeFragmentParam(key, null, "replace");
      }
    }
    if (!controller || Object.keys(bindings).length === 0) {
      this.windowRestoration.set(entry.name, { disposers: [], keys });
      return;
    }

    const params = fragmentParams();
    let restored = false;
    this.applyingFragmentState = true;
    try {
      for (const [key, path] of Object.entries(bindings)) {
        const scopedKey = scopedFragmentKey(entry.name, key);
        if (!params.has(scopedKey)) continue;
        writePath(controller, path.split("."), params.get(scopedKey));
        restored = true;
      }
    } finally {
      this.applyingFragmentState = false;
    }
    controller.controllerDidRestore({ restored });

    const disposers: Array<() => void> = [];
    for (const [key, path] of Object.entries(bindings)) {
      const scopedKey = scopedFragmentKey(entry.name, key);
      const segments = path.split(".");
      disposers.push(
        observePath(controller, segments, () => {
          if (this.applyingFragmentState) return;
          this.writeFragmentParam(scopedKey, fragmentValue(readPath(controller, segments)), "replace");
        }),
      );
    }
    this.windowRestoration.set(entry.name, { disposers, keys });
  }

  private finishWindowActivation(): void {
    for (const entry of this.windows) finishActivation(entry.root);
  }

  private disposeWindowRestoration(windowName: string): Set<string> {
    const existing = this.windowRestoration.get(windowName);
    if (!existing) return new Set();
    for (const dispose of existing.disposers) dispose();
    this.windowRestoration.delete(windowName);
    return existing.keys;
  }

  private installFragmentRouting(): void {
    if ((globalThis as any).__fragmentRoutingInstalled) return;
    (globalThis as any).__fragmentRoutingInstalled = true;
    const apply = (): void => {
      this.applyFragmentToWindows();
    };
    window.addEventListener("popstate", apply);
    window.addEventListener("hashchange", apply);
  }

  private applyFragmentToWindows(): void {
    const params = fragmentParams();
    for (const entry of [...this.windows]) {
      if (!entry.name) continue;
      const requestedContent = params.get(entry.name)?.trim() || null;
      if (requestedContent && requestedContent !== entry.contentName && this.windowContainer(entry.name)) {
        if (this.hasWindowContent(requestedContent)) {
          void this.loadWindowContent(entry.name, requestedContent, { history: "none" });
        } else {
          console.warn(`[Swill] window="${entry.name}" requested unknown content "${requestedContent}"`);
        }
        continue;
      }
      this.restoreWindowState(entry, { pruneStale: false, writeContent: false });
    }
  }

  private writeFragmentParam(key: string, value: string | null, mode: FragmentHistoryMode): void {
    if (mode === "none" || this.applyingFragmentState) return;
    writeFragmentParam(key, value, mode);
  }

  // ---- document-level listeners ----

  private installFocusTracking(): void {
    if ((globalThis as any).__focusTrackingInstalled) return;
    (globalThis as any).__focusTrackingInstalled = true;
    // focusin fires on every DOM focus change inside the document,
    // including the one synthesized by Tab returning from the browser URL
    // bar.
    document.addEventListener(
      "focusin",
      (e) => {
        const fe = e as FocusEvent;
        syncFirstResponderFromFocus(fe.target as HTMLElement | null, fe.relatedTarget as HTMLElement | null);
      },
      { capture: true },
    );
    // focusout with relatedTarget === null means focus is leaving the
    // document entirely (Tab to URL bar, click outside, etc.). Clear the
    // cached first responder so the next focusin starts clean.
    document.addEventListener(
      "focusout",
      (e) => {
        if ((e as FocusEvent).relatedTarget === null) {
          syncFirstResponderFromFocus(null);
        }
      },
      { capture: true },
    );
  }

  private installKeyRouting(): void {
    if ((globalThis as any).__keyRoutingInstalled) return;
    (globalThis as any).__keyRoutingInstalled = true;
    document.addEventListener("keydown", (event) => {
      const target = (event.target as HTMLElement | null) ?? (document.activeElement as HTMLElement | null);
      if (!target) return;
      firstResponderFor(target)?.keyDown(event);
    });
    document.addEventListener("keyup", (event) => {
      const target = (event.target as HTMLElement | null) ?? (document.activeElement as HTMLElement | null);
      if (!target) return;
      firstResponderFor(target)?.keyUp(event);
    });
  }

  private watchForChanges(root: HTMLElement): void {
    const observer = new MutationObserver((records) => {
      const added: HTMLElement[] = [];

      for (const rec of records) {
        rec.removedNodes.forEach((n) => {
          if (!(n instanceof HTMLElement)) return;
          for (const view of topViewsIn(n)) detachViewTree(view);
        });
        rec.addedNodes.forEach((n) => {
          if (n instanceof HTMLElement) added.push(n);
        });
      }

      if (added.length) {
        for (const el of added) wireSubtree(el);
      }
    });
    observer.observe(root, { childList: true, subtree: true });
  }

  private installConsoleHooks(): void {
    const root = this.rootElement;
    Object.defineProperty(globalThis, "__controllers", {
      configurable: true,
      get: () => collectControllers(root),
    });
    Object.defineProperty(globalThis, "__root", {
      configurable: true,
      get: () => collectControllers(root).find((c) => !c.parent) ?? null,
    });
    Object.defineProperty(globalThis, "__controllerFor", {
      configurable: true,
      value: (el: HTMLElement | string): Controller | null => {
        if (typeof el === "string") {
          const found = (root.querySelector(el) ?? document.querySelector(el)) as HTMLElement | null;
          return found ? controllerFor(found) : null;
        }
        return controllerFor(el);
      },
    });
    Object.defineProperty(globalThis, "__firstResponder", {
      configurable: true,
      get: () => currentFirstResponder(),
    });
    Object.defineProperty(globalThis, "__application", {
      configurable: true,
      get: () => this,
    });
  }
}

File: opal/lib/swill/view/application.rb
# backtick_javascript: true

module Swill
  class Application < Responder
    class << self
      attr_writer :shared

      def shared
        @shared ||= new
      end
    end

    def start(root = `document.body`)
      self.class.shared = self
      # The application is the chain top — the responder chain bottoms out here
      # and the first responder falls back here. Set before wiring so
      # after_load can make a view the first responder.
      FirstResponder.chain_top = self
      @root = root
      @window_templates = {}
      @captured_window_contents = {}
      @windows = []
      @fragments = FragmentRouter.new
      scan_window_templates
      prepare_window_containers(root)
      before_launch
      controllers = Awakening.wire(root, appear: false)
      register_window_containers(root)
      @windows.each { |window| window.restore_state(@fragments, prune_stale: false, write_content: true) }
      Awakening.activate(controllers)
      watch(root)
      install_event_listeners(root)
      @fragments.on_change { apply_fragment_to_windows }
      after_launch
      self
    end

    # Cocoa puts this transition on NSWindow. Until first-responder state
    # moves onto Swill::Window, the application owns it. Explicit requests
    # honor the responder's policy gate before entering the first-responder
    # state machine.
    def make_first_responder(responder)
      return false if responder && !responder.accepts_first_responder?

      FirstResponder.make(responder)
    end

    # Present a template-backed dialog. Returns a Promise that resolves when
    # the window is dismissed.
    def show_window(name, into: `document.body`)
      saved_first_responder = FirstResponder.current
      controller = instantiate_window(name.to_s, into)
      root = controller.view.element

      `#{root}.show && #{root}.show()`

      FirstResponder.make(controller)

      promise = Promise.new
      @windows << Window.new(
        name: name,
        root: root,
        controller: controller,
        saved_first_responder: saved_first_responder,
        close_resolver: -> { promise.resolve(nil) }
      )
      promise
    end

    def load_window_content(window_name, content_name, history: :push)
      window = window_named(window_name.to_s) || register_container(window_container(window_name.to_s))
      raise ArgumentError, %(Application: no window container "#{window_name}") unless window

      container = window.root
      previous = FirstResponder.current
      detach_window_content(container)
      `#{container}.replaceChildren()`

      node = clone_window_content(content_name.to_s)
      `#{container}.appendChild(#{node})`
      controllers = Awakening.wire(container, appear: false)
      window.controller = top_controller_in(container, controllers)
      window.content_name = content_name.to_s

      `#{container}.setAttribute("name", #{content_name.to_s})`
      @fragments.write(window.name, content_name.to_s, history)
      window.restore_state(@fragments, prune_stale: history != :none, write_content: history != :none)
      Awakening.activate(controllers)
      if window.controller
        FirstResponder.make(window.controller)
      elsif previous && Focus.responder_element(previous) && `#{Focus.responder_element(previous)}.parentElement`
        FirstResponder.make(previous)
      end

      window.controller
    end

    def dismiss(controller)
      window = @windows.find { |candidate| candidate.contains_controller?(controller) }
      return unless window

      @windows.delete(window)
      window.dismiss
    end

    def next_responder
      nil
    end

    # Launch lifecycle, Cocoa nouns in Sequel hook structure:
    #   before_launch — before the root subtree is awakened.
    #   after_launch  — the subtree is wired and observed; Cocoa's
    #                   applicationDidFinishLaunching moment.
    def before_launch; end
    def after_launch; end

    private

    # ---- windows ----

    # Named [window] containers only — dialogs share the array but are never
    # targets for load_window_content.
    def window_named(name)
      @windows.find { |window| window.container? && window.name == name }
    end

    def window_for_root(root)
      @windows.find { |window| window.root?(root) }
    end

    def register_window_containers(root)
      window_containers(root).each do |container|
        register_container(container) unless window_for_root(container)
      end
    end

    def register_container(container)
      return nil unless container

      name = element_attribute(container, "window")
      return nil unless name

      window = Window.new(
        name: name,
        root: container,
        content_name: element_attribute(container, "name"),
        controller: top_controller_in(container, Awakening.controllers_within(container))
      )
      @windows << window
      window
    end

    def window_containers(root)
      containers = []
      containers << root if `#{root}.hasAttribute && #{root}.hasAttribute("window")`
      `Array.from(#{root}.querySelectorAll("[window]"))`.each { |container| containers << container }
      containers
    end

    def window_container(name)
      window_containers(@root).find { |container| element_attribute(container, "window") == name }
    end

    def detach_window_content(container)
      Awakening.controllers_within(container).each do |controller|
        next if `#{controller.view.element} === #{container}`

        Awakening.detach(controller.view.element)
      end
    end

    def top_controller_in(container, controllers)
      controllers.find do |controller|
        element = controller.view.element
        parent = controller.parent
        `#{container}.contains(#{element})` &&
          (parent.nil? || !`#{container}.contains(#{parent.view.element})`)
      end
    end

    def invalidate_restorable_state(controller)
      window = @windows.find { |candidate| candidate.controller.equal?(controller) }
      window&.write_current_state(@fragments)
    end

    # ---- fragment routing ----

    def apply_fragment_to_windows
      params = @fragments.params
      @windows.dup.each do |window|
        requested = params[window.name]
        if requested && requested != window.content_name
          if window_content?(requested)
            load_window_content(window.name, requested, history: :none)
          else
            warn(%([Swill] window="#{window.name}" requested unknown content "#{requested}"))
          end
        else
          window.restore_state(@fragments, prune_stale: false, write_content: false)
        end
      end
    end

    # ---- window templates and content ----

    def scan_window_templates
      @window_templates ||= {}
      `Array.from(document.querySelectorAll("template"))`.each do |template|
        name = `#{template}.getAttribute("name")`
        next unless name

        for_window = `#{template}.getAttribute("for") === "window"`
        body_child = `#{template}.parentElement === document.body`
        @window_templates[name.to_s] = template if for_window || body_child
      end
      @window_templates
    end

    # Look up a window template, rescanning the document once on a miss so
    # templates inserted after launch are still found.
    def window_template(name)
      @window_templates[name] || scan_window_templates[name]
    end

    def window_content?(name)
      return false unless name

      name = name.to_s
      return true if @captured_window_contents.key?(name)

      # The template is a raw DOM node; test presence on the JS side.
      `!!#{window_template(name)}`
    end

    def prepare_window_containers(root)
      window_containers(root).each do |container|
        window_name = element_attribute(container, "window")
        default_name = element_attribute(container, "name") || window_name
        if `#{container}.children.length > 0`
          first = `#{container}.firstElementChild`
          @captured_window_contents[default_name.to_s] ||= `#{first}.cloneNode(true)` if first && default_name
        end

        requested = @fragments.params[window_name.to_s]
        content_name = requested || default_name
        unless window_content?(content_name)
          warn(%([Swill] window="#{window_name}" requested unknown content "#{content_name}")) if requested
          content_name = default_name
        end
        next unless content_name
        next if `#{container}.children.length > 0` && content_name.to_s == default_name.to_s

        `#{container}.replaceChildren()`
        `#{container}.appendChild(#{clone_window_content(content_name.to_s)})`
        `#{container}.setAttribute("name", #{content_name.to_s})`
      end
    end

    def instantiate_window(name, into)
      node = clone_window_content(name)
      raise ArgumentError, %(Application: empty window template "#{name}") unless node

      `#{into}.appendChild(#{node})`
      controllers = Awakening.wire(node)
      controller = Awakening.controller_for(node) || controllers.first
      raise ArgumentError, %(Application: window "#{name}" root has no controller attribute) unless controller

      controller
    end

    def clone_window_content(name)
      captured = @captured_window_contents[name]
      return `#{captured}.cloneNode(true)` if captured

      template = window_template(name)
      raise ArgumentError, %(Application: no window content template "#{name}") unless template

      `const content = #{template}.content;
       const node = content && content.firstElementChild
         ? content.firstElementChild
         : #{template}.children[0];
       return node ? node.cloneNode(true) : null;`
    end

    def element_attribute(element, name)
      value = `#{element}.getAttribute(#{name})`
      value && value.to_s
    end

    # ---- document-level event routing ----

    # Reconcile the first responder with DOM focus changes, and route key
    # events to the current first responder. focusin/focusout bubble to the
    # root; keydown/keyup bubble up from the focused element.
    def install_event_listeners(root)
      focus_in = ->(event) { Focus.sync_from_focus(`#{event}.target`) }
      `#{root}.addEventListener("focusin", #{focus_in})`

      focus_out = lambda do |event|
        related = `#{event}.relatedTarget`
        # Only reconcile when focus leaves to nowhere; a move to another element
        # is handled by that element's focusin.
        Focus.sync_from_focus(nil, `#{event}.target`) if `#{related} == null`
      end
      `#{root}.addEventListener("focusout", #{focus_out})`

      key_down = ->(event) { FirstResponder.current&.key_down(Event.new(event)) }
      `#{root}.addEventListener("keydown", #{key_down})`

      key_up = ->(event) { FirstResponder.current&.key_up(Event.new(event)) }
      `#{root}.addEventListener("keyup", #{key_up})`
    end

    def watch(root)
      callback = lambda do |mutations, _observer|
        mutations.each do |mutation|
          `Array.from(#{mutation}.removedNodes)`.each do |node|
            Awakening.detach(node) if `#{node}.nodeType === 1`
          end
          `Array.from(#{mutation}.addedNodes)`.each do |node|
            Awakening.wire(node) if `#{node}.nodeType === 1`
          end
        end
      end

      @observer = `new MutationObserver(#{callback})`
      `#{@observer}.observe(#{root}, { childList: true, subtree: true })`
    end
  end
end
